File: lib/iriq/shape.rb
module Iriq
  # Structured route shape: an ordered list of typed segment entries plus
  # rendering methods that produce the various string forms (placeholder,
  # canonical-dates, raw-types, etc.).
  #
  # Replaces the string-as-data convention where PathShape's String output
  # was the only carrier of shape information. Structured Shape makes:
  #   - downstream consumers cheap (they iterate entries instead of
  #     re-deriving from segments + classifier)
  #   - shape identity explicit (structural #== / #hash, not string match)
  #   - multiple renderings free (canonical dates, hints on/off, raw types
  #     vs hinted) without re-walking segments
  #
  # The cluster identity layer still uses string keys for storage; a
  # follow-up step migrates Cluster equality to be Shape-driven.
  class Shape
    attr_reader :entries

    # Build a Shape from raw path segments using the given classifier.
    def self.from_segments(segments, classifier: SegmentClassifier::DEFAULT)
      new(entries: SegmentHints.derive(segments || [], classifier))
    end

    # Build a Shape from already-derived SegmentHints entries — same input
    # PathShape.from_entries used to take. Useful when the caller already
    # walked segments once and wants to avoid a second pass.
    def self.from_entries(entries)
      new(entries: entries || [])
    end

    def initialize(entries:)
      @entries = entries
    end

    def empty?
      @entries.empty?
    end

    # Render to the placeholder form — "/users/{user_id}" etc. This is the
    # default string representation.
    def render(hints: true, canonical_dates: false, canonical_currencies: false)
      return "/" if empty?

      "/" + @entries.map { |e|
        render_entry(e, hints: hints, canonical_dates: canonical_dates,
                        canonical_currencies: canonical_currencies)
      }.join("/")
    end

    def to_s
      render
    end
    alias inspect to_s

    # Structural equality: two Shapes are equal when they render the same
    # placeholder form. /users/1 and /users/999 are the same shape even
    # though raw values differ, but /users/1 and /posts/1 are not.
    def ==(other)
      other.is_a?(Shape) && other.render == render
    end
    alias eql? ==

    def hash
      render.hash
    end

    def to_dump
      { "entries" => @entries.map { |e| e.transform_keys(&:to_s) } }
    end

    def self.from_dump(h)
      entries = (h["entries"] || []).map do |e|
        e.each_with_object({}) do |(k, v), acc|
          key = k.to_sym
          # Only :type is symbolized — :value and :hint stay as strings,
          # matching what SegmentHints.derive produces.
          acc[key] = key == :type ? v.to_sym : v
        end
      end
      new(entries: entries)
    end

    private

    def render_entry(entry, hints:, canonical_dates:, canonical_currencies:)
      return entry[:value] unless entry[:variable]

      if canonical_dates && entry[:type] == :date &&
         (canon = SegmentClassifier.canonical_date(entry[:value]))
        return canon
      end

      if canonical_currencies && entry[:type] == :currency &&
         (canon = SegmentClassifier.canonical_currency(entry[:value]))
        return canon
      end

      placeholder = if hints
                      entry[:hint] || SegmentClassifier.display_type(entry[:type])
                    else
                      SegmentClassifier.display_type(entry[:type])
                    end
      "{#{placeholder}}"
    end
  end
end

File: rust/iriq/src/shape.rs
use crate::classifier::{
    canonical_currency, canonical_date, display_type, SegmentClassifier, SegmentType,
    DEFAULT_CLASSIFIER,
};
use crate::hints::{derive_hints, SegmentHint};

#[derive(Debug, Clone, Default)]
pub struct Shape {
    pub entries: Vec<SegmentHint>,
}

#[derive(Debug, Clone, Copy, Default)]
pub struct ShapeRenderOptions {
    /// Disables hint placeholders; when true, variable segments render as
    /// their display type (e.g. `{integer}`) rather than `{user_id}`.
    pub hints_off: bool,
    pub canonical_dates: bool,
    pub canonical_currencies: bool,
}

impl Shape {
    pub fn from_segments(segments: &[String], classifier: Option<&SegmentClassifier>) -> Self {
        let c = classifier.unwrap_or(&DEFAULT_CLASSIFIER);
        Shape {
            entries: derive_hints(segments, c),
        }
    }

    pub fn from_entries(entries: Vec<SegmentHint>) -> Self {
        Shape { entries }
    }

    pub fn render(&self, opts: ShapeRenderOptions) -> String {
        if self.entries.is_empty() {
            return "/".to_string();
        }
        let mut s = String::new();
        for e in &self.entries {
            s.push('/');
            s.push_str(&render_entry(e, opts));
        }
        s
    }

    pub fn equal(&self, other: &Shape) -> bool {
        self.render(ShapeRenderOptions::default()) == other.render(ShapeRenderOptions::default())
    }
}

impl std::fmt::Display for Shape {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.render(ShapeRenderOptions::default()))
    }
}

fn render_entry(e: &SegmentHint, opts: ShapeRenderOptions) -> String {
    if !e.variable {
        return e.value.clone();
    }
    if opts.canonical_dates && e.ty == SegmentType::Date {
        if let Some(c) = canonical_date(&e.value) {
            return c;
        }
    }
    if opts.canonical_currencies && e.ty == SegmentType::Currency {
        if let Some(c) = canonical_currency(&e.value) {
            return c;
        }
    }
    if !opts.hints_off && !e.hint.is_empty() {
        return format!("{{{}}}", e.hint);
    }
    format!("{{{}}}", display_type(e.ty))
}
